modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Classes.js

```diff
@@ -0,0 +1,53 @@
+// Enhances common node/mark types to accept a class parameter
+// and filter out classes that don't match the list on paste/parse
+import { Extension } from '@tiptap/core';
+export default (options) => {
+  // Create a class allowlist map for each element
+  const allow = {};
+  options.styles.forEach(style => {
+    const tag = style.tag.toLowerCase();
+    allow[tag] = (allow[tag] || []).concat(...(style.class ? style.class.split(' ') : []));
+  });
+  return Extension.create({
+    addGlobalAttributes() {
+      return [
+        {
+          types: Object.keys(options.types),
+          attributes: {
+            class: {
+              default: null,
+              renderHTML(attributes) {
+                return {
+                  class: attributes.class
+                };
+              },
+              parseHTML(element) {
+                const tag = element.tagName.toLowerCase();
+                // This tag is not configured
+                if (!allow[tag]) {
+                  return {
+                    class: null
+                  };
+                }
+                const classes = (element.getAttribute('class') || '')
+                  .split(' ')
+                  .filter(c => allow[tag].includes(c));
+                // If we have valid classes, join and return them.
```

Comment:
If "h1" has two choices for "styles", one with a class and one without, will it get stuck with the class every time because of the "default to first" rule?

Reply:
If the editor is actually typing into the RTE and using our style control to apply styles, `parseHTML`, where the best guess default class is applied, is not triggered and `default: null` does the correct thing.

I am making a small enhancement to keep track of tags passed in without classes (null) so that the case you outlined works correctly when text is pasted in from an external source.